[chore] add update version automation workflow for artifacts release

jackgopack4 opened the pull request:
Add quality-of-life enhancement to automate bumping version number in the Makefile as well as the various distribution manifests (as the number of distros grows, this step will become more of a pain).

Feel free to run in your own fork to verify; also see the successful run in my fork: 
https://github.com/jackgopack4/opentelemetry-collector-releases/actions/runs/12320554731/job/34390067049
As well as the generated PR in my fork:
https://github.com/jackgopack4/opentelemetry-collector-releases/pull/23

If this gets merged I will open a PR to update collector release doc in `opentelemetry-collector` repo.

Automatic version bump (increase minor version number) is default behavior; this can be overridden by providing a manual version number update (like bugfix v0.110.0 -> v0.110.1)

andrzej-stencel reviewed (commented):
I :100: agree that bumping the version should be a simple, single step. This change implements this as a series of steps in GitHub Actions, which makes it not usable if you want to run it outside of GitHub Actions. I'd rather see a shell script written that does the version updates, and that script called as a single step in GitHub Actions. Committing changes should be separate from the script, so that anyone can run the script locally and preview the changes before making it a commit.

Note that there are (to my knowledge) three potentially different versions that need to be bumped in the manifests, as described [here](https://github.com/open-telemetry/opentelemetry-collector/issues/11129#issuecomment-2343499207):

> - stable components all use the stable version - currently v1.15.0
> - unstable components from core repository use the core's version - currently v0.109.0
> - unstable components from contrib repository use the contrib's version - currently v0.109.0
> 
> The core and contrib versions are usually the same, but they are not guaranteed to be the same.

jackgopack4 commented:
> I 💯 agree that bumping the version should be a simple, single step. This change implements this as a series of steps in GitHub Actions, which makes it not usable if you want to run it outside of GitHub Actions. I'd rather see a shell script written that does the version updates, and that script called as a single step in GitHub Actions. Committing changes should be separate from the script, so that anyone can run the script locally and preview the changes before making it a commit.

Would be happy to do it this way, just was modeling it after the existing action script in collector core: https://github.com/open-telemetry/opentelemetry-collector/blob/main/.github/workflows/prepare-release.yml

> Note that there are (to my knowledge) three potentially different versions that need to be bumped in the manifests, as described [here](https://github.com/open-telemetry/opentelemetry-collector/issues/11129#issuecomment-2343499207):
> 
> > * stable components all use the stable version - currently v1.15.0
> > * unstable components from core repository use the core's version - currently v0.109.0
> > * unstable components from contrib repository use the contrib's version - currently v0.109.0

Got it, will add stable and beta component update regardless of the method of execution

jackgopack4 commented:
@andrzej-stencel made the changes you requested; runs in one bash script that can be ran locally in addition to in github actions and adds stable/beta functionality

andrzej-stencel commented on .github/workflows/scripts/bump-versions.sh:
```diff
@@ -0,0 +1,144 @@
+
+echo "Version update completed."
+
+Commit changes and draft PR
```
As mentioned in https://github.com/open-telemetry/opentelemetry-collector-releases/pull/684#pullrequestreview-2345254646:

> Committing changes should be separate from the script, so that anyone can run the script locally and preview the changes before making it a commit.

At the very least, the commit, push, PR creation part should be opt-in via a script option like `--commit` or similar. The default (when the `--commit` option is not passed to the script) should be to not make the commit.

jackgopack4 replied:
added two flags, --commit which commits changes and creates new branch, pushing to remote, and --pull-request which creates draft PR with that new branch. Github workflow passes --commit every time, but allows release manager to check or uncheck --pull-request (checked by default).

andrzej-stencel commented on .github/workflows/update-version.yaml:
```diff
@@ -0,0 +1,47 @@
+      - name: Run bump-versions.sh
+        run: |
+          chmod +x .github/workflows/scripts/bump-versions.sh
+          .github/workflows/scripts/bump-versions.sh \
+            --current_beta "${{ github.event.inputs.current_beta }}" \
+            --current_stable "${{ github.event.inputs.current_stable }}" \
+            --next_beta "${{ github.event.inputs.next_beta }}" \
+            --next_stable "${{ github.event.inputs.next_stable }}"
```
As mentioned in https://github.com/open-telemetry/opentelemetry-collector-releases/pull/684#pullrequestreview-2345254646, there are three distinct versions used in the builder configurations. Two of those are equal _most of the time_. This script assumes that core and contrib versions are always the same, which means it will do the wrong thing in case there's a mismatch between core and contrib versions.

I'm reluctant to have an automation in place that works correctly _most of the time_, as opposed to _always_. Releasing software should be as frictionless as possible, given that it is sometimes done under pressure (e.g. for hot fix releases).

I propose to update the script to accept three versions, and perform the replacements separately for core and contrib components. It is possible to easily tell those apart as the module paths are different.

jackgopack4 replied:
apologies, I must not have read your comment thoroughly enough. I agree with your proposed change and will push an update to the PR branch shortly.

andrzej-stencel commented on .github/workflows/scripts/bump-versions.sh:
```diff
@@ -0,0 +1,144 @@
+  "distributions/otelcol-k8s/manifest.yaml"
+  "distributions/otelcol-otlp/manifest.yaml"
+  "Makefile"
+  # "cmd/builder/builder-config-yaml"
```
This can be removed, as https://github.com/open-telemetry/opentelemetry-collector-releases/pull/671/ does not add this file anymore.

andrzej-stencel commented on .github/workflows/scripts/bump-versions.sh:
```diff
@@ -0,0 +1,144 @@
+
+# Function to display usage
+usage() {
+  echo "Usage: $0 --current_beta <current_beta> [--current_stable <current_stable>] [--next_beta <next_beta>] [--next_stable <next_stable>]"
```
This is a nit, but words in option names should probably be separated with a dash `-` and not an underscore `_`.
```suggestion
  echo "Usage: $0 --current-beta <current_beta> [--current-stable <current_stable>] [--next-beta <next_beta>] [--next-stable <next_stable>]"
```

jackgopack4 replied:
fixed, thanks

andrzej-stencel commented on .github/workflows/scripts/bump-versions.sh:
```diff
@@ -0,0 +1,144 @@
+for file in "${files[@]}"; do
+  if [ -f "$file" ]; then
+    if [ -n "$current_beta" ]; then
+      sed -i.bak "s/$current_beta/$next_beta/g" "$file"
```
We don't need the `g` option, right? The version only appears once on each line.
```suggestion
      sed -i.bak "s/$current_beta/$next_beta/" "$file"
```

jackgopack4 replied:
fixed, thanks

andrzej-stencel commented on .github/workflows/scripts/bump-versions.sh:
```diff
@@ -0,0 +1,210 @@
+      line_number=$((line_number + 1))
+    done < "$file"
+    mv "$temp_file" "$file"
+    rm "${temp_file}.bak"
```
The script logs errors trying to remove non-existent `.bak` files when run with only the `--current-beta-contrib` flag:

```console
./.github/workflows/scripts/bump-versions.sh --current-beta-contrib v0.111.0
rm: cannot remove '/tmp/tmp.h6AQVOKLTG.bak': No such file or directory
rm: cannot remove '/tmp/tmp.VsQcTga32N.bak': No such file or directory
Version update completed.
Changes not committed and PR not created.
```

I propose to add an existence check before removing the `.bak` files.

```suggestion
    [ -f "${temp_file}.bak" ] && rm "${temp_file}.bak"
```

jackgopack4 replied:
got rid of the backup logic so this problem should be solved

andrzej-stencel commented on .github/workflows/scripts/bump-versions.sh:
```diff
@@ -0,0 +1,210 @@
+    temp_file=$(mktemp)
+    cp "$file" "$temp_file"
+    line_number=1
+    while IFS= read -r line; do
```
Why does the script run `sed` for each line separately? Putting efficiency aside, this seems harder to understand than just running `sed` on the whole file(s). The replacement patterns are also less precise. For example, there are lines in `Makefile` that have `github.com/open-telemetry/opentelemetry-collector-contrib` in them, but are not subject to replacements.

I would imagine running the `sed` command on the whole files and specifying the replacement patterns specifically for each purpose. For example:
- to replace core versions in manifests: `sed "s/\(^.*go\.opentelemetry\.io\/collector\/.*\) v0\.111\.0/\1 v0.112.0/" distributions/otelcol/manifest.yaml`
- to replace `otelcol_version` in manifests: `sed "s/\(^\s\+otelcol_version:\) 0\.111\.0/\1 0.112.0/" distributions/otelcol/manifest.yaml`
- etc.

andrzej-stencel commented on .github/workflows/scripts/bump-versions.sh:
```diff
@@ -0,0 +1,210 @@
+# Parse named arguments
+while [[ "$#" -gt 0 ]]; do
+  case $1 in
+    --current-beta-core) current_beta_core="$2"; shift ;;
```
Thinking about this more, do we actually even need the "current" versions from the user? Seems that we can read the current versions from the manifest files, e.g. with `awk`:

```console
$ current_beta_core=$(awk '/^.*go\.opentelemetry\.io\/collector\/.* v0/ {print $4; exit}' distributions/otelcol/manifest.yaml)
$ echo $current_beta_core
v0.111.0
```

andrzej-stencel reviewed (commented):
The script in the current form does not update the following places:

- `otelcol_version` and `version` in `manifest.yaml` files
- `OTELCOL_BUILDER_VERSION` in `Makefile`

I think the script should use the `--next-beta-core` version for the `OTELCOL_BUILDER_VERSION` in `Makefile` and for the `otelcol_version` field in manifests. As for the `version` field in manifest, it's probably max(next_beta_core, next_beta_contrib). Or we could add a new parameter to the script and action to specify it.

jackgopack4 commented:
successful release PR in my fork: https://github.com/jackgopack4/opentelemetry-collector-releases/pull/21

jackgopack4 commented:
@andrzej-stencel @mx-psi waiting for checks to run but I feel better about this attempt. incorporated feedback from code review and added changelog step

jackgopack4 commented:
> The script in the current form does not update the following places:
> 
> * `otelcol_version` and `version` in `manifest.yaml` files
> * `OTELCOL_BUILDER_VERSION` in `Makefile`
> 
> I think the script should use the `--next-beta-core` version for the `OTELCOL_BUILDER_VERSION` in `Makefile` and for the `otelcol_version` field in manifests. As for the `version` field in manifest, it's probably max(next_beta_core, next_beta_contrib). Or we could add a new parameter to the script and action to specify it.

addressed both of these

jackgopack4 commented on CHANGELOG.md:
required to make this work. Can remove from this PR once https://github.com/open-telemetry/opentelemetry-collector-releases/pull/775 is merged

jackgopack4 replied:
yes, done

andrzej-stencel reviewed (commented):
Thank you for your patience with my comments @jackgopack4. Just a couple small things and this will be ready to merge.

jackgopack4 commented on .github/workflows/update-version.yaml:
```diff
@@ -0,0 +1,43 @@
+
+      - name: Run bump-versions.sh
+        run: |
+          chmod +x .github/workflows/scripts/bump-versions.sh && \
```
yes I didn't realize git makes it executable by default; fixed

andrzej-stencel reviewed (approved):
Thanks Jack!
